File: gtx-out/src/CardDemo.OnlineTransactionManagement.Application/AddTransactionResult.cs
namespace CardDemo.OnlineTransactionManagement.Application;

/// <summary>Result of adding a new transaction.</summary>
public sealed record AddTransactionResult
{
    /// <summary>Whether the operation succeeded.</summary>
    public required bool Success { get; init; }

    /// <summary>Error code if the operation failed.</summary>
    public TransactionError Error { get; init; }

    /// <summary>The generated transaction ID on success.</summary>
    public long TransactionId { get; init; }

    /// <summary>Updated account balance after the transaction.</summary>
    public decimal UpdatedBalance { get; init; }
}

File: gtx-out/src/main/java/com/grapeup/carddemo/onlinetransactionmanagement/application/AddTransactionResult.java
package com.grapeup.carddemo.onlinetransactionmanagement.application;

import java.math.BigDecimal;
import java.util.Optional;

/**
 * Result of the add-transaction use case.
 * On success, contains the generated transaction ID and updated balance.
 * On failure, contains the error code.
 */
public record AddTransactionResult(
        boolean success,
        String transactionId,
        BigDecimal updatedBalance,
        TransactionError error
) {
    public static AddTransactionResult ok(String transactionId, BigDecimal updatedBalance) {
        return new AddTransactionResult(true, transactionId, updatedBalance, null);
    }

    public static AddTransactionResult fail(TransactionError error) {
        return new AddTransactionResult(false, null, null, error);
    }
}
